### PR TITLE
Draft for removing blob contents from ISnapshotTree.blobs

### PR DESCRIPTION
**Pull request by jatgarg**
Refer: https://github.com/microsoft/FluidFramework/issues/4746
This is just a draft and want to take some hints/suggestions on the problem as @vladsud suggested.
1.) This Pr achieves removing blobId -> blob Contents mapping from the snapshot taken in detached container
(serialize) for draft mode.
2.) Lower layers like datastores/dds can use storage to read blob contents from cached blobs.
3.) One main problem is that we do need to read blob contents synchronously while de-serializing the container. So I had to supply the blob contents but for that I added blobContents in a separate property in ISnapshotTreeWithBlobContents and read from that.

### CODE REVIEW

File: packages/loader/container-loader/src/utils.ts [hunk 1]
```diff
@@ -6,7 +6,13 @@
 import { parse } from "url";
 import { v4 as uuid } from "uuid";
 import { assert, bufferToString, stringToBuffer } from "@fluidframework/common-utils";
-import { ISummaryTree, ISnapshotTree, SummaryType } from "@fluidframework/protocol-definitions";
+import { ISnapshotTree, ISummaryTree, SummaryType } from "@fluidframework/protocol-definitions";
+
+// This will got in common place to loader and runtime as it used by both.
+export interface ISnapshotTreeWithBlobContents extends ISnapshotTree {
```
Comment (vladsud) [inline packages/loader/container-loader/src/utils.ts:12]: This interface formalizes contract and reduces chance of extra conversions (based64 and back) which is good thing.
But does usage of this interface passes through compat boundary, or runtime and loader have independent usage? I'd guess that convertSnapshotTreeToSummaryTree() is the primary consumer if it's the later, but isn't usage of ContainerStorageAdapter() ensures that runtime does not need to know about it (at least in first approximation, there is a need to revisit sync -> async changes when loading local data stores).

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -1387,29 +1381,20 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
-                ]);
-            } else {
-                members = readAndParseFromBlobs<[string, ISequencedClient][]>(snapshot.trees[".protocol"].blobs,
-                    baseTree.blobs.quorumMembers);
-                proposals = readAndParseFromBlobs<[number, ISequencedProposal, string[]][]>(
-                    snapshot.trees[".protocol"].blobs, baseTree.blobs.quorumProposals);
-                values = readAndParseFromBlobs<[string, ICommittedProposal][]>(snapshot.trees[".protocol"].blobs,
-                    baseTree.blobs.quorumValues);
-            }
+            [members, proposals, values] = await Promise.all([
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1393]: Irrespective of other changes, I really like this unification and random code not needing to know these differences. Consider doing it as separate PR, hiding these differences in readAndParse (i.e. allowing undefined storage as a staging solution).

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 3]
```diff
@@ -845,24 +845,35 @@ export class LocalFluidDataStoreContextBase extends FluidDataStoreContext {
...
+        let attributes;
         if (snapshot !== undefined) {
-            // Note: storage can be undefined in special case while detached.
-            const attributes = this.storage !== undefined
-                ? await readAndParse<ReadFluidDataStoreAttributes>(
-                    this.storage, snapshot.blobs[dataStoreAttributesBlobName])
-                : readAndParseFromBlobs<ReadFluidDataStoreAttributes>(
-                    snapshot.blobs, snapshot.blobs[dataStoreAttributesBlobName]);
-
+            // Need to rip through snapshot.
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:851]: It feels like there is an opportunity to reuse code, as I'd think other implementations of context would have similar logic

**Comment by vladsud** [comment]
One thing to clarify: I assume we have no solution (in this draft PR) to a problem of delay loaded local data stores / DDSs in detached container and current need to create summary in a synchronous way on container attachment, right? In other words, until we have some story on how container attachment can be solved, there is no path forward. Is that right?

**Comment by jatgarg** [comment]
> One thing to clarify: I assume we have no solution (in this draft PR) to a problem of delay loaded local data stores / DDSs in detached container and current need to create summary in a synchronous way on container attachment, right? In other words, until we have some story on how container attachment can be solved, there is no path forward. Is that right?

The solution is to have the format used in here which has blob contents in different property in the snapshot. So that whill attaching, we can read contents from that as I doing in this PR.

**Comment by vladsud** [comment]
Got it (I forgot about it :)). I think it's good starting point (as it makes things more explicit, which is always a good thing), but in principle it does not address the problem of relying on existing mechanism (storage interface) for blob retrieval until we can solve the problem of attached container relying on synchronous process.
Attachment of data stores / dds in attached container also has this dependency, but I think it's much easier to solve.

**Comment by jatgarg** [comment]
Closing as this has been merged with smaller changes in all the above mentioned PRs.

**Pull request closed without merging**